### ISSUE TITLE
Iterable protocol should be implemented for `__getitem__`

### ISSUE THREAD

**Issue by AleCandido** [microsoft/pyright#4307]
**Describe the bug**
Iterable not correctly detected.

**To Reproduce**
Define a class with just `__getitem__`, and try to call `list()` on its instance.

**Expected behavior**
No error should be reported.

**Screenshots or Code**
```py
from dataclasses import dataclass

@dataclass
class Container:
    a: int
    b: float

    def __getitem__(self, key): -> Any
        return getattr(self, "ab"[key])

list(Container(1, 3.0))  # flagged by PyRight
```

while the following is not flagged:
```py
list(iter(Container(1, 3.0)))
```

**VS Code extension or command-line**
On Neovim through [`coc-pyright`](https://github.com/fannheyward/coc-pyright) plugin (should be equivalent to VS code), version 1.1.283.

**Additional context**
This issue was already raised by #3702 (or at least is extremely similar to that one), and declared solved by v1.1.261.

**Comment by erictraut** [comment]
Pyright's behavior is correct here. Not surprisingly, mypy emits the same error in this case. The `list` constructor is defined as accepting an `Iterable`, and `Container` does not meet that protocol definition. If you think the constructor definition for `list` is incorrect, please report the problem to the maintainers of the [typeshed project](https://www.google.com/search?q=typeshed+github&oq=typeshed+github&aqs=edge.0.69i59j69i60l3.5152j0j1&sourceid=chrome&ie=UTF-8). Pyright (like mypy and other type checkers) get their type definitions for stdlib from typeshed.

**Comment by AleCandido** [comment]
Ok, thank you for your answer. It is true that is not `Iterable` since it does not expose an `__iter__` method, but at runtime `list` is actually using `__getitem__` and builds.

For the sake of the `typeshed` report, can you confirm that this is the definition used for the `list` constructor?
https://github.com/python/typeshed/blob/a74df38df0c74f45cd1ba9ab7b0d22e561bc531b/stdlib/collections/__init__.pyi#L97

**Comment by erictraut** [comment]
The definition is in `builtins.pyi` on [this line](https://github.com/python/typeshed/blob/a74df38df0c74f45cd1ba9ab7b0d22e561bc531b/stdlib/builtins.pyi#L962).

**Comment by AleCandido** [comment]
Ok thank you, I got the wrong one.

**Comment by erictraut** [comment]
If you're able to modify the code in question, another option is to simply implement the `Iterable` protocol in `Container` (i.e. `__iter__` and `__next__`) rather than `__getitem__`.

**Comment by AleCandido** [comment]
In the specific case I can, but `__getitem__` is the one I mainly want (and much simpler to implement), and becoming a `list` is a secondary benefit.

However, an iterable is also an object implementing only a suitable `__getitem__` method (with "sequence semantics"), according to the [reference](https://docs.python.org/3/glossary.html#term-iterable).
But I can get that this is impossible to check during static analysis, so an `Iterable` has to implement a proper `__iter__`.

Maybe is not an error at all after all, it is simply a part of the reference that is not possible to reflect in type checking tools.
